File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/util/SimpleStack.java
package com.apollographql.apollo.internal.util;

import java.util.ArrayList;
import java.util.List;

/**
 * Simple stack data structure which accepts null elements. Backed by list.
 * @param <E>
 */
public class SimpleStack<E> {

  private List<E> backing;

  public SimpleStack() {
    backing = new ArrayList<>();
  }

  public SimpleStack(int initialSize) {
    backing = new ArrayList<>(initialSize);
  }

  public void push(E element) {
    backing.add(element);
  }

  public E pop() {
    if (isEmpty()) {
      throw new IllegalStateException("Stack is empty.");
    }
    return backing.remove(backing.size() - 1);
  }

  public boolean isEmpty() {
    return backing.isEmpty();
  }
}

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/SimpleStack.kt
package com.apollographql.apollo.cache.normalized.internal

/**
 * Simple stack data structure which accepts null elements. Backed by list.
 * @param <E>
</E> */
class SimpleStack<E> {
  private val backing: MutableList<E>

  constructor() {
    backing = ArrayList()
  }

  constructor(initialSize: Int) {
    backing = ArrayList(initialSize)
  }

  fun push(element: E) {
    backing.add(element)
  }

  fun pop(): E {
    check(!isEmpty) { "Stack is empty." }
    return backing.removeAt(backing.size - 1)
  }

  val isEmpty: Boolean
    get() = backing.isEmpty()
}